From 133005d4560b5066d9150d616bfbb8e4d80f17a6 Mon Sep 17 00:00:00 2001
From: David Teigland <teigland@redhat.com>
Date: Mon, 3 Nov 2025 15:44:13 -0600
Subject: [PATCH 17/32] vgsplit: update PR on devs to match destination VG

Stop or start PR on PVs being moved to match the PR
settings of the destination VG.

(cherry picked from commit 987aab1e26e129e2969a30aaa0c475a395a55b22)
---
 lib/device/persist.c | 22 ++++++++++++++++++++++
 lib/device/persist.h |  1 +
 tools/vgsplit.c      | 18 ++++++++++++++++++
 3 files changed, 41 insertions(+)

diff --git a/lib/device/persist.c b/lib/device/persist.c
index 21ab839d0..2b22c8f0a 100644
--- a/lib/device/persist.c
+++ b/lib/device/persist.c
@@ -1682,6 +1682,28 @@ void persist_finish_after(struct cmd_context *cmd, struct volume_group *vg, stru
 	persist_key_file_remove(cmd, vg);
 }
 
+int persist_stop_devs(struct cmd_context *cmd, struct volume_group *vg, struct dm_list *stop_devs)
+{
+	char *local_key = (char *)find_config_tree_str(cmd, local_pr_key_CFG, NULL);
+	int local_host_id = find_config_tree_int(cmd, local_host_id_CFG, NULL);
+	char our_key_buf[PR_KEY_BUF_SIZE] = { 0 };
+	uint64_t our_key_val = 0;
+
+	if (vg_is_sanlock(vg))
+		local_key = NULL;
+
+	if (!local_key && !local_host_id)
+		return 1;
+
+	if (!get_our_key(cmd, vg, local_key, local_host_id, our_key_buf, &our_key_val))
+		return_0;
+
+	if (!_run_stop(cmd, vg, stop_devs, our_key_buf, 0))
+		return_0;
+
+	return 1;
+}
+
 int persist_stop(struct cmd_context *cmd, struct volume_group *vg)
 {
 	DM_LIST_INIT(devs);
diff --git a/lib/device/persist.h b/lib/device/persist.h
index d94fd1edd..3792ee580 100644
--- a/lib/device/persist.h
+++ b/lib/device/persist.h
@@ -49,6 +49,7 @@ int persist_start(struct cmd_context *cmd, struct volume_group *vg,
 		  const char *update_our_key);
 
 int persist_stop(struct cmd_context *cmd, struct volume_group *vg);
+int persist_stop_devs(struct cmd_context *cmd, struct volume_group *vg, struct dm_list *stop_devs);
 
 int persist_finish_before(struct cmd_context *cmd, struct volume_group *vg, struct dm_list *devs, char **key);
 void persist_finish_after(struct cmd_context *cmd, struct volume_group *vg, struct dm_list *devs, char *key);
diff --git a/tools/vgsplit.c b/tools/vgsplit.c
index 2d93c1935..8397d4176 100644
--- a/tools/vgsplit.c
+++ b/tools/vgsplit.c
@@ -14,6 +14,7 @@
  */
 
 #include "tools.h"
+#include "lib/device/persist.h"
 
 static int _lv_is_in_vg(struct volume_group *vg, struct logical_volume *lv)
 {
@@ -525,6 +526,7 @@ int vgsplit(struct cmd_context *cmd, int argc, char **argv)
 	struct volume_group *vg_to, *vg_from = NULL;
 	DM_LIST_INIT(devs_moved);
 	struct device *dev_moved;
+	int pr_from, pr_to;
 	int opt;
 	int existing_vg = 0;
 	int r = ECMD_FAILED;
@@ -711,6 +713,19 @@ int vgsplit(struct cmd_context *cmd, int argc, char **argv)
 	/* store it on disks */
 	log_verbose("Writing out updated volume groups");
 
+	/*
+	 * Make PR of moved PVs match the VG they are moved to.
+	 * !pr_from && pr_to: on dest, not on source: start dest PR before move
+	 * pr_from && !pr_to: on source, not on dest: stop dest PR after move
+	 */
+	pr_from = (vg_from->pr & (VG_PR_REQUIRE|VG_PR_AUTOSTART)) ? 1 : 0;
+	pr_to = (vg_to->pr & (VG_PR_REQUIRE|VG_PR_AUTOSTART)) ? 1 : 0;
+
+	if (!pr_from && pr_to && !persist_start(cmd, vg_to,
+						(char *)find_config_tree_str(cmd, local_pr_key_CFG, NULL),
+						find_config_tree_int(cmd, local_host_id_CFG, NULL), NULL, NULL))
+		goto_bad;
+
 	/*
 	 * First, write out the new VG as EXPORTED.  We do this first in case
 	 * there is a crash - we will still have the new VG information, in an
@@ -755,6 +770,9 @@ int vgsplit(struct cmd_context *cmd, int argc, char **argv)
 
 	backup(vg_to);
 
+	if (pr_from && !pr_to && !persist_stop_devs(cmd, vg_to, &devs_moved))
+		log_warn("WARNING: Failed to stop PR.");
+
 	log_print_unless_silent("%s volume group \"%s\" successfully split from \"%s\"",
 				existing_vg ? "Existing" : "New",
 				vg_to->name, vg_from->name);
-- 
2.52.0

